### PR TITLE
Integrate ProgressManager with Python Lab

### PR DESCRIPTION
**Pull request by molly-moen**
Note: this PR should be merged after [this pythonlab-packages](https://github.com/code-dot-org/pythonlab-packages/pull/5) PR is in.

This PR integrates the lab2 progress system with Python Lab. Python Lab validation is a bit different from Music Lab validation. In Music Lab, validations are defined in json per-level, and are conditions that we check via custom, engineer-owned code while the user's program runs. In Python Lab, there is a validation unittest file that is run when the "test" button is clicked by users, and we want all their tests to pass in order for them to pass the level.

The conditions are sent as part of the level properties and utilized by the progress container to update the progress manager. I wanted to be able to re-use this existing plumbing, so I made a new wrapper method on `level.rb`, `get_validations`, which usually returns `properties['validation']`. However, in the case of Python Lab it checks if the level has a validation file, and if it does it returns the hard-coded condition of all tests passed (otherwise it returns nil).

I then created a `PythonValidator`, which takes a `PythonValidationTracker` as a parameter, which is a class with a singleton. This singleton is used when we run tests to record the test results. Then when we check progress via the Progress Manager, the validator checks if we have the all tests passed condition, and if we do we check if all the tests passed (if the tests were run and there were no tests, we count that as a pass for now).

The only other change that impacts all lab2 labs is we now reset the ProgressManager on both `levelValidations` change and `levelId` change, since all Python Lab levels have 1 of two `levelValidations` (null or all passed), but we still want to rest the progress between levels.

After these changes, the progress bubble updates appropriately for levels with validation!

## Links
- This is a part of: [CT-526](https://codedotorg.atlassian.net/browse/CT-526). I will do a follow-up with the actual title of the ticket, gating the continue/finish button on the status of the validation. This PR felt big enough. 

## Testing story
Tested locally and added unit tests. I also verified that Music lab validation still works in English and Spanish.

## Follow-up work
See links.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/levels/level.rb [hunk 1]
```diff
@@ -874,7 +874,7 @@ def project_type
   # Whether this level has validation for the completion of student work.
   def validated?
     if uses_lab2?
-      return properties.dig('level_data', 'validations').present?
```
Comment (molly-moen) [inline dashboard/app/models/levels/level.rb:877]: Discussed [on slack](https://codedotorg.slack.com/archives/C044AGTKW3D/p1723657092363209) that this was an incorrect check after some changes by @sanchitmalhotra126. Accessing `properties['validations']` (what `get_validations` does) is the correct change.

File: dashboard/app/models/levels/pythonlab.rb [hunk 2]
```diff
@@ -78,4 +77,24 @@ def clean_up_predict_settings
...
+
+  # Return the validation condition for this level. If the level has a validation file, the condition
+  # is that all tests passed. If there is no validation file, there are no conditions.
+  def get_validations
+    has_validation = start_sources && start_sources["files"]&.any? {|(_, file)| file["type"] == 'validation'}
+    if has_validation
+      [{
+        conditions: [
+          {
+            name: 'PASSED_ALL_TESTS',
```
Comment (sanchitmalhotra126) [inline dashboard/app/models/levels/pythonlab.rb:89]: Out of curiosity, are there plans to eventually move this into levelbuilder with multiple conditions (ex. `PASSED_FIRST_TEST`, `PASSED_SOME_TESTS`, etc)?

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

File: dashboard/app/models/levels/pythonlab.rb (hunk 2 above)
Comment (molly-moen) [reply dashboard/app/models/levels/pythonlab.rb:89]: Not currently planned, I think we will always want them to pass all tests.

**Pull request merged**